### ISSUE TITLE
Curl Post a file of exactly 16384 bytes

### ISSUE THREAD

**Issue by tjpowell** [curl/curl#2889]
I have code that continually posts files to a web address.  The code looks like the following:

	form = curl_mime_init(curl);
						field = curl_mime_addpart(form);
						curl_mime_name(field, "cvdfile");
						curl_mime_filedata(field,  file_name.c_str());

						curl_easy_setopt(curl, CURLOPT_POST, 1L);

						std::string buildstr = "http://";
						buildstr += local_file_stores_it->server;
						buildstr += ":";
						buildstr += local_file_stores_it->port;
						buildstr += "/api/writefile";

						curl_easy_setopt(curl, CURLOPT_URL, buildstr);

						std::string ret_string;
						ret_string = "";
						curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, CurlWrite_CallbackFunc_StdString);
						curl_easy_setopt(curl, CURLOPT_WRITEDATA, &ret_string);

						INFO("%dDVR: curl post size is %ld", local_thread_number, file_info.st_size);

                                               //had to put this code in to fix it
						if (file_info.st_size == 16384){
							file_info.st_size =  16383;
							INFO("%dDVR: curl post size is now %ld", local_thread_number, file_info.st_size);
						}
						curl_easy_setopt(curl, CURLOPT_POSTFIELDSIZE, file_info.st_size);
						curl_easy_setopt(curl, CURLOPT_MIMEPOST, form);
						curl_easy_setopt(curl, CURLOPT_CONNECTTIMEOUT, 30);
						
						res = curl_easy_perform(curl);
						/* Check for errors */ 
						if(res != CURLE_OK){
							FATAL("%dDVR:curl_easy_perform() failed: %s",local_thread_number,
								  curl_easy_strerror(res));
						}

### I expected the following
I happened to have a file that was EXACTLY 16384 bytes and it wouldn't post, giving me this error message:

35008DVR:curl_easy_perform() failed: Failure when receiving data from the peer.

file_info.st_size would be 16384.
This would happen every time.  Also when this happens, it appears curl does not release the windows lock on the file...I can't move the file to another directory until after I kill this process.

As a workaround, I tried seeing what would happen if I put in 16383 instead of 16384, and the file posts successfully.

Hunting around the header files, I see #define CURL_MAX_WRITE_SIZE 16384 so I'm wondering if its some sort of edge case where the size happens to be the exact size of this particular buffer, it doesn't work.

### curl/libcurl version
curl 7.61.  I was on curl 7.59 then upgraded yesterday hoping to solve this problem, but it persisted.
[curl -V output]

### operating system

Windows server 2016.

**Comment by bagder** [comment]
Why do you set `CURLOPT_POSTFIELDSIZE` at all and why that value? You're doing a formbased upload so the actual body will be many bytes larger than the exact file size.

**Comment by tjpowell** [comment]
I’m doing it the size of the file.   I coded this awhile ago and got it to work.    It looks like you have a point there tho….
However I would say its still interesting that it gets this error with this particular value still.

-Tim

**Comment by bagder** [comment]
> However I would say its still interesting that it gets this error with this particular value still.

I agree, but it could also just be the receiving server acting differently since neither sending 16384 nor 16383 will not send the full file and thus be an incomplete post so I it's not my top-prio to understand.

**Comment by bagder** [comment]
I just tested `CURLOPT_POSTFIELDSIZE` on a formpost and yes, it simply makes a too short post that is less data than the `Content-Length:` said up front. It doesn't cause a problem in itself for me though. Tested both against a local Apache and with nc. Cannot reproduce.

**Comment by tjpowell** [comment]
The file I’m uploading always makes it there in full.  This is for a DVR and the files have I am uploading vs what was received are shown to be identical.

I’ve also discovered it’s any multiple of 16384 that causes the same issue.  I have another file that was exactly 163,840 that exhibited the same symptoms.

I had changed my “fix” to:

if ((file_info.st_size % 16384) == 0){
                                                file_info.st_size -=  1;
                                                FATAL("%dDVR: curl post size is now %ld, shouldn't have to do this", local_thread_number, file_info.st_size);
}
curl_easy_setopt(curl, CURLOPT_POSTFIELDSIZE, file_info.st_size);

And this seems to work for me without any loss of data.

I’m a bit hazy about what exactly all the fields should be as far as forms and mime and types, since web gets/posts isn’t my primary expertise so I had just fiddled around with it till it worked.
I would guess that if I had a file size of says 16184 and the form was 200 for a total of 16384,  I would see the same issue.

-Tim

**Comment by tjpowell** [comment]
Taking out

curl_easy_setopt(curl, CURLOPT_POSTFIELDSIZE, file_info.st_size);

altogether seems to work for me for the 16kb files.

I made sure I saw the error first, took out that line, then it worked.  Double checked to make sure I was sane.

I’m always a fan of removing lines of code to solve a problem.  I like this solution far better than the hack I was putting in.

Thanks.

-Tim
